src/megaapi_impl.cpp

```diff
@@ -16938,42 +16938,24 @@ MegaNodeList *MegaApiImpl::getChildren(MegaNode* p, int order)
         return new MegaNodeListPrivate();
     }
 
-    sdkMutex.lock();
-    Node *parent = client->nodebyhandle(p->getHandle());
-    if (!parent || parent->type == FILENODE)
-    {
-        sdkMutex.unlock();
-        return new MegaNodeListPrivate();
-    }
-
     node_vector childrenNodes;
 
-    if (std::function<bool(Node*, Node*)> comparatorFunction = getComparatorFunction(order, *client))
+    SdkMutexGuard guard(sdkMutex);
+
+    Node *parent = client->nodebyhandle(p->getHandle());
+    if (parent && parent->type != FILENODE)
     {
+        childrenNodes.reserve(parent->children.size());
         for (node_list::iterator it = parent->children.begin(); it != parent->children.end(); )
         {
-            Node *n = *it++;
-            const node_vector::iterator i = std::lower_bound(childrenNodes.begin(), childrenNodes.end(), n, comparatorFunction);
-            childrenNodes.insert(i, n);
-        }
-    }
-    else
-    {
-        for (node_list::iterator it = parent->children.begin(); it != parent->children.end(); )
             childrenNodes.push_back(*it++);
+        }
+        if (std::function<bool(Node*, Node*)> comparatorFunction = getComparatorFunction(order, *client))
+        {
+            std::sort(childrenNodes.begin(), childrenNodes.end(), comparatorFunction);
+        }
     }
-
-    MegaNodeListPrivate *result = NULL;
-    if (childrenNodes.size())
-    {
-        result = new MegaNodeListPrivate(childrenNodes.data(), int(childrenNodes.size()));
-    }
-    else
-    {
-        result = new MegaNodeListPrivate();
-    }
-    sdkMutex.unlock();
-    return result;
+    return new MegaNodeListPrivate(childrenNodes.data(), int(childrenNodes.size()));;
```

Comment:
We could take this chance to apply the same improvements to the other analog methods:
- `MegaApiImpl::getFileFolderChildren()`
- `MegaApiImpl::getIndex()`
The other public methods that allows to specify an order for the results already use `ORDER_NONE` by default (ie. `search()` ) :)

Reply:
I agree but I'd like to wait and see if the change from lower_bound to sort() has any unintended issues - in theory I think the old lower_bound technique would produce something similar to stable_sort, but with equal ranges reversed.    I think we should keep that in mind for future improvements once we are certain this one is good.    And getIndex is an interesting case, I'm not sure about that one since, just before it locks the mutex or just after is releases it, the child list could have changed, and then the caller may get back an incorrect or out of date index - if they use that in a subsequent call to get that item, they will usually get the right one but occasionally get something else.

Reply:
The `getIndex()` looked weird/risky to me also, I don't see why the app would need to ask the SDK about the index of an already returned sorted list of children. Apparently, it's used only by Windows Phone 8, which is currently obsolete and is replaced by the Universal Windows Platform. Nevertheless, out of curiosity, I've checked the code and it's used right after addition of a new node. The app wants to know the index, in the list of children, to insert it in the UI in the right position.
It's subject to races, as you pointed out. I think we should tag the method as deprecated.

Reply:
Ok, I've made the changes here - getFileFolderChildren is quite a lot smaller!   Since I was making changes there anyway, I've arranged it in such a way that the new allocations will behave nicely if an out-of-memory error is thrown, passing in unique_ptr to the list contstructors, which meant a few small changes in the list class, just using unique_ptr instead of raw. 
About getIndex, that's good news that it's not used anymore.  Let's just remove it then, and avoid it being used accidentally.   If there are any complaints we can explain the issue, and if necessary provide an alternative (maybe they pass us the list they know, and we can provide the index against that list).    I think for ourselves, it's better to remove the bad code than still have it there where it's a maintenance issue.  thanks!